File: tests/algorithms/transpiler/FS/sorts/external_sort.fs
// Generated 2025-08-11 16:20 +0700

exception Return
let mutable _nowSeed:int64 = 0L
let mutable _nowSeeded = false
let _initNow () =
    let s = System.Environment.GetEnvironmentVariable("MOCHI_NOW_SEED")
    if System.String.IsNullOrEmpty(s) |> not then
        match System.Int32.TryParse(s) with
        | true, v ->
            _nowSeed <- int64 v
            _nowSeeded <- true
        | _ -> ()
let _now () =
    if _nowSeeded then
        _nowSeed <- (_nowSeed * 1664525L + 1013904223L) % 2147483647L
        int _nowSeed
    else
        int (System.DateTime.UtcNow.Ticks % 2147483647L)

_initNow()
let _idx (arr:'a array) (i:int) : 'a =
    if not (obj.ReferenceEquals(arr, null)) && i >= 0 && i < arr.Length then arr.[i] else Unchecked.defaultof<'a>
let _arrset (arr:'a array) (i:int) (v:'a) : 'a array =
    let mutable a = arr
    if i >= a.Length then
        let na = Array.zeroCreate<'a> (i + 1)
        Array.blit a 0 na 0 a.Length
        a <- na
    a.[i] <- v
    a
let _repr v =
    let s = sprintf "%A" v
    s.Replace("[|", "[")
     .Replace("|]", "]")
     .Replace("; ", ", ")
let _floordiv (a:int) (b:int) : int =
    let q = a / b
    let r = a % b
    if r <> 0 && ((a < 0) <> (b < 0)) then q - 1 else q
let rec subarray (xs: int array) (start: int) (``end``: int) =
    let mutable __ret : int array = Unchecked.defaultof<int array>
    let mutable xs = xs
    let mutable start = start
    let mutable ``end`` = ``end``
    try
        let mutable result: int array = Array.empty<int>
        let mutable k: int = start
        while k < ``end`` do
            result <- Array.append result [|(_idx xs (int k))|]
            k <- k + 1
        __ret <- result
        raise Return
        __ret
    with
        | Return -> __ret
and merge (left_half: int array) (right_half: int array) =
    let mutable __ret : int array = Unchecked.defaultof<int array>
    let mutable left_half = left_half
    let mutable right_half = right_half
    try
        let mutable result: int array = Array.empty<int>
        let mutable i: int = 0
        let mutable j: int = 0
        while (i < (Seq.length (left_half))) && (j < (Seq.length (right_half))) do
            if (_idx left_half (int i)) < (_idx right_half (int j)) then
                result <- Array.append result [|(_idx left_half (int i))|]
                i <- i + 1
            else
                result <- Array.append result [|(_idx right_half (int j))|]
                j <- j + 1
        while i < (Seq.length (left_half)) do
            result <- Array.append result [|(_idx left_half (int i))|]
            i <- i + 1
        while j < (Seq.length (right_half)) do
            result <- Array.append result [|(_idx right_half (int j))|]
            j <- j + 1
        __ret <- result
        raise Return
        __ret
    with
        | Return -> __ret
and merge_sort (array: int array) =
    let mutable __ret : int array = Unchecked.defaultof<int array>
    let mutable array = array
    try
        if (Seq.length (array)) <= 1 then
            __ret <- array
            raise Return
        let middle: int = _floordiv (Seq.length (array)) 2
        let left_half: int array = subarray (array) (0) (middle)
        let right_half: int array = subarray (array) (middle) (Seq.length (array))
        let sorted_left: int array = merge_sort (left_half)
        let sorted_right: int array = merge_sort (right_half)
        __ret <- merge (sorted_left) (sorted_right)
        raise Return
        __ret
    with
        | Return -> __ret
and split_into_blocks (data: int array) (block_size: int) =
    let mutable __ret : int array array = Unchecked.defaultof<int array array>
    let mutable data = data
    let mutable block_size = block_size
    try
        let mutable blocks: int array array = Array.empty<int array>
        let mutable i: int = 0
        while i < (Seq.length (data)) do
            let ``end``: int = if (i + block_size) < (Seq.length (data)) then (i + block_size) else (Seq.length (data))
            let block: int array = subarray (data) (i) (``end``)
            let sorted_block: int array = merge_sort (block)
            blocks <- Array.append blocks [|sorted_block|]
            i <- ``end``
        __ret <- blocks
        raise Return
        __ret
    with
        | Return -> __ret
and merge_blocks (blocks: int array array) =
    let mutable __ret : int array = Unchecked.defaultof<int array>
    let mutable blocks = blocks
    try
        let num_blocks: int = Seq.length (blocks)
        let mutable indices: int array = Array.empty<int>
        let mutable i: int = 0
        while i < num_blocks do
            indices <- Array.append indices [|0|]
            i <- i + 1
        let mutable result: int array = Array.empty<int>
        let mutable ``done``: bool = false
        while not ``done`` do
            ``done`` <- true
            let mutable min_val: int = 0
            let mutable min_block: int = 0 - 1
            let mutable j: int = 0
            while j < num_blocks do
                let idx: int = _idx indices (int j)
                if idx < (Seq.length (_idx blocks (int j))) then
                    let ``val``: int = _idx (_idx blocks (int j)) (int idx)
                    if (min_block = (0 - 1)) || (``val`` < min_val) then
                        min_val <- ``val``
                        min_block <- j
                    ``done`` <- false
                j <- j + 1
            if not ``done`` then
                result <- Array.append result [|min_val|]
                indices.[int min_block] <- (_idx indices (int min_block)) + 1
        __ret <- result
        raise Return
        __ret
    with
        | Return -> __ret
and external_sort (data: int array) (block_size: int) =
    let mutable __ret : int array = Unchecked.defaultof<int array>
    let mutable data = data
    let mutable block_size = block_size
    try
        let mutable blocks: int array array = split_into_blocks (data) (block_size)
        __ret <- merge_blocks (blocks)
        raise Return
        __ret
    with
        | Return -> __ret
and main () =
    let mutable __ret : unit = Unchecked.defaultof<unit>
    try
        let __bench_start = _now()
        let __mem_start = System.GC.GetTotalMemory(true)
        let data: int array = unbox<int array> [|7; 1; 5; 3; 9; 2; 6; 4; 8; 0|]
        let sorted_data: int array = external_sort (data) (3)
        printfn "%s" (_repr (sorted_data))
        let __bench_end = _now()
        let __mem_end = System.GC.GetTotalMemory(true)
        printfn "{\n  \"duration_us\": %d,\n  \"memory_bytes\": %d,\n  \"name\": \"main\"\n}" ((__bench_end - __bench_start) / 1000) (__mem_end - __mem_start)

        __ret
    with
        | Return -> __ret
main()

File: tests/algorithms/x/PHP/sorts/external_sort.php
<?php
ini_set('memory_limit', '-1');
function _append($arr, $x) {
    $arr[] = $x;
    return $arr;
}
function _intdiv($a, $b) {
    if ($b === 0 || $b === '0') {
        throw new DivisionByZeroError();
    }
    if (function_exists('bcdiv')) {
        $sa = is_int($a) ? strval($a) : (is_string($a) ? $a : sprintf('%.0f', $a));
        $sb = is_int($b) ? strval($b) : (is_string($b) ? $b : sprintf('%.0f', $b));
        return intval(bcdiv($sa, $sb, 0));
    }
    return intdiv($a, $b);
}
function _iadd($a, $b) {
    if (function_exists('bcadd')) {
        $sa = is_int($a) ? strval($a) : (is_string($a) ? $a : sprintf('%.0f', $a));
        $sb = is_int($b) ? strval($b) : (is_string($b) ? $b : sprintf('%.0f', $b));
        return bcadd($sa, $sb, 0);
    }
    return $a + $b;
}
function _isub($a, $b) {
    if (function_exists('bcsub')) {
        $sa = is_int($a) ? strval($a) : (is_string($a) ? $a : sprintf('%.0f', $a));
        $sb = is_int($b) ? strval($b) : (is_string($b) ? $b : sprintf('%.0f', $b));
        return bcsub($sa, $sb, 0);
    }
    return $a - $b;
}
function _imul($a, $b) {
    if (function_exists('bcmul')) {
        $sa = is_int($a) ? strval($a) : (is_string($a) ? $a : sprintf('%.0f', $a));
        $sb = is_int($b) ? strval($b) : (is_string($b) ? $b : sprintf('%.0f', $b));
        return bcmul($sa, $sb, 0);
    }
    return $a * $b;
}
function _idiv($a, $b) {
    return _intdiv($a, $b);
}
function _imod($a, $b) {
    if (function_exists('bcmod')) {
        $sa = is_int($a) ? strval($a) : (is_string($a) ? $a : sprintf('%.0f', $a));
        $sb = is_int($b) ? strval($b) : (is_string($b) ? $b : sprintf('%.0f', $b));
        return intval(bcmod($sa, $sb));
    }
    return $a % $b;
}
function subarray($xs, $start, $end) {
  $result = [];
  $k = $start;
  while ($k < $end) {
  $result = _append($result, $xs[$k]);
  $k = _iadd($k, 1);
};
  return $result;
}
function merge($left_half, $right_half) {
  $result = [];
  $i = 0;
  $j = 0;
  while ($i < count($left_half) && $j < count($right_half)) {
  if ($left_half[$i] < $right_half[$j]) {
  $result = _append($result, $left_half[$i]);
  $i = _iadd($i, 1);
} else {
  $result = _append($result, $right_half[$j]);
  $j = _iadd($j, 1);
}
};
  while ($i < count($left_half)) {
  $result = _append($result, $left_half[$i]);
  $i = _iadd($i, 1);
};
  while ($j < count($right_half)) {
  $result = _append($result, $right_half[$j]);
  $j = _iadd($j, 1);
};
  return $result;
}
function merge_sort($array) {
  if (count($array) <= 1) {
  return $array;
}
  $middle = _idiv(count($array), 2);
  $left_half = subarray($array, 0, $middle);
  $right_half = subarray($array, $middle, count($array));
  $sorted_left = merge_sort($left_half);
  $sorted_right = merge_sort($right_half);
  return merge($sorted_left, $sorted_right);
}
function split_into_blocks($data, $block_size) {
  $blocks = [];
  $i = 0;
  while ($i < count($data)) {
  $end = (_iadd($i, $block_size) < count($data) ? _iadd($i, $block_size) : count($data));
  $block = subarray($data, $i, $end);
  $sorted_block = merge_sort($block);
  $blocks = _append($blocks, $sorted_block);
  $i = $end;
};
  return $blocks;
}
function merge_blocks($blocks) {
  $num_blocks = count($blocks);
  $indices = [];
  $i = 0;
  while ($i < $num_blocks) {
  $indices = _append($indices, 0);
  $i = _iadd($i, 1);
};
  $result = [];
  $done = false;
  while (!$done) {
  $done = true;
  $min_val = 0;
  $min_block = _isub(0, 1);
  $j = 0;
  while ($j < $num_blocks) {
  $idx = $indices[$j];
  if ($idx < count($blocks[$j])) {
  $val = $blocks[$j][$idx];
  if ($min_block == (_isub(0, 1)) || $val < $min_val) {
  $min_val = $val;
  $min_block = $j;
};
  $done = false;
}
  $j = _iadd($j, 1);
};
  if (!$done) {
  $result = _append($result, $min_val);
  $indices[$min_block] = _iadd($indices[$min_block], 1);
}
};
  return $result;
}
function external_sort($data, $block_size) {
  $blocks = split_into_blocks($data, $block_size);
  return merge_blocks($blocks);
}
function main() {
  $data = [7, 1, 5, 3, 9, 2, 6, 4, 8, 0];
  $sorted_data = external_sort($data, 3);
  echo str_replace('false', 'False', str_replace('true', 'True', str_replace('"', '\'', str_replace(':', ': ', str_replace(',', ', ', json_encode($sorted_data, 1344)))))), PHP_EOL;
}
main();
